<!---
Latex Macros
-->
$$
\newcommand{\bar}{\,|\,}
\newcommand{\Xs}{\mathcal{X}}
\newcommand{\Ys}{\mathcal{Y}}
\newcommand{\y}{\mathbf{y}}
\newcommand{\weights}{\mathbf{w}}
\newcommand{\balpha}{\boldsymbol{\alpha}}
\newcommand{\bbeta}{\boldsymbol{\beta}}
\newcommand{\aligns}{\mathbf{a}}
\newcommand{\align}{a}
\newcommand{\source}{\mathbf{s}}
\newcommand{\target}{\mathbf{t}}
\newcommand{\ssource}{s}
\newcommand{\starget}{t}
\newcommand{\repr}{\mathbf{f}}
\newcommand{\repry}{\mathbf{g}}
\newcommand{\x}{\mathbf{x}}
\newcommand{\prob}{p}
\newcommand{\vocab}{V}
\newcommand{\params}{\boldsymbol{\theta}}
\newcommand{\param}{\theta}
\DeclareMathOperator{\perplexity}{PP}
\DeclareMathOperator{\argmax}{argmax}
\DeclareMathOperator{\argmin}{argmin}
\newcommand{\train}{\mathcal{D}}
\newcommand{\counts}[2]{\#_{#1}(#2) }
\newcommand{\length}[1]{\text{length}(#1) }
\newcommand{\indi}{\mathbb{I}}
$$

# Assignment 3

## Introduction

In the last assignment, you will apply deep learning methods to solve a particular story understanding problem. Automatic understanding of stories is an important task in natural language understanding [[1]](http://anthology.aclweb.org/D/D13/D13-1020.pdf). Specifically, you will develop a model that given a sequence of sentences learns to sort these sentence in order to yield a coherent story [[2]](https://www.microsoft.com/en-us/research/wp-content/uploads/2016/06/short-commonsense-stories.pdf). This sounds (and to an extent is) trivial for humans, however it is a quite difficult task for machines as it involves commonsense knowledge and temporal understanding.

## Goal

You are given a dataset of 45502 instances, each consisting of 5 sentences. Your system needs to ouput a sequence of numbers which represent the predicted order of these sentences. For example, given a story:

    He went to the store.
    He found a lamp he liked.
    He bought the lamp.
    Jan decided to get a new lamp.
    Jan's lamp broke.

your system needs to provide an answer in the following form:

    2	3	4	1	0

where the numbers correspond to the zero-based index of each sentence in the correctly ordered story. So "`2`" for "`He went to the store.`" means that this sentence should come 3rd in the correctly ordered target story. In This particular example, this order of indices corresponds to the following target story:

    Jan's lamp broke.
    Jan decided to get a new lamp.
    He went to the store.
    He found a lamp he liked.
    He bought the lamp.

## Resources

To develop your model(s), we provide a training and a development datasets. The test dataset will be held out, and we will use it to evaluate your models. The test set is coming from the same task distribution, and you don't need to expect drastic changes in it.

You will use [TensorFlow](https://www.tensorflow.org/) to build a deep learning model for the task. We provide a very crude system which solves the task with a low accuracy, and a set of additional functions you will have to use to save and load the model you create so that we can run it.

As we have to run the notebooks of each submission, and as deep learning models take long time to train, your notebook **NEEDS** to conform to the following requirements:
* You **NEED** to run your parameter optimisation offline, and provide your final model saved by using the provided function
* The maximum size of a zip file you can upload to moodle is 160MB. We will **NOT** allow submissions larger than that.
* We do not have time to train your models from scratch! You **NEED** to provide the full code you used for the training of your model, but by all means you **CANNOT** call the training method in the notebook you will send to us.
* We will run these notebooks automatically. If your notebook runs the training procedure, in addition to loading the model, and we need to edit your code to stop the training, you will be penalised with **-20 points**.
* If you do not provide a pretrained model, and rely on training your model on our machines, you will get **0 points**.
* It needs to be tested on the stat-nlp-book Docker setup to ensure that it does not have any dependencies outside of those that we provide. If your submission fails to adhere to this requirement, you will get **0 points**.

Running time and memory issues:
* We have tested a possible solution on a mid-2014 MacBook Pro, and a few epochs of the model run in less than 3min. Thus it is possible to train a model on the data in reasonable time. However, be aware that you will need to run these models many times over, for a larger number of epochs (more elaborate models, trained on much larger datasets can train for weeks! However, this shouldn't be the case here.). If you find training times too long for your development cycle you can reduce the training set size. Once you have found a good solution you can increase the size again. Caveat: model parameters tuned on a smaller dataset may not be optimal for a larger training set.
* In addition to this, as your submission is capped by size, feel free to experiment with different model sizes, numeric values of different precisions, filtering the vocabulary size, downscaling some vectors, etc.

## Hints

A non-exhaustive list of things you might want to give a try:
- better tokenization
- experiment with pre-trained word representations such as [word2vec](https://code.google.com/archive/p/word2vec/), or [GloVe](http://nlp.stanford.edu/projects/glove/). Be aware that these representations might take a lot of parameters in your model. Be sure you use only the words you expect in the training/dev set and account for OOV words. When saving the model parameters, pre-rained word embeddings can simply be used in the word embedding matrix of your model. As said, make sure that this word embedding matrix does not contain all of word2vec or GloVe. Your submission is limited, and we will not allow uploading nor using the whole representations set (up to 3GB!)
- reduced sizes of word representations
- bucketing and batching (our implementation is deliberately not a good one!)
  - make sure to draw random batches from the data! (we do not provide this in our code!)
- better models:
  - stacked RNNs (see tf.nn.rnn_cell.MultiRNNCel
  - bi-directional RNNs
  - attention
  - word-by-word attention
  - conditional encoding
  - get model inspirations from papers on nlp.stanford.edu/projects/snli/
  - sequence-to-sequence encoder-decode architecture for producing the right ordering
- better training procedure:
  - different training algorithms
  - dropout on the input and output embeddings (see tf.nn.dropout)
  - L2 regularization (see tf.nn.l2_loss)
  - gradient clipping (see tf.clip_by_value or tf.clip_by_norm)
- model selection:
  - early stopping
- hyper-parameter optimization (e.g. random search or grid search (expensive!))
    - initial learning rate
    - dropout probability
    - input and output size
    - L2 regularization
    - gradient clipping value
    - batch size
    - ...
- post-processing
  - for incorporating consistency constraints

## Setup Instructions
It is important that this file is placed in the **correct directory**. It will not run otherwise. The correct directory is

    DIRECTORY_OF_YOUR_BOOK/assignments/2016/assignment3/problem/group_X/
    
where `DIRECTORY_OF_YOUR_BOOK` is a placeholder for the directory you downloaded the book to, and in `X` in `group_X` contains the number of your group.

After you placed it there, **rename the notebook file** to `group_X`.

The notebook is pre-set to save models in

    DIRECTORY_OF_YOUR_BOOK/assignments/2016/assignment3/problem/group_X/model/

Be sure not to tinker with that - we expect your submission to contain a `model` subdirectory with a single saved model! 
The saving procedure might overwrite the latest save, or not. Make sure you understand what it does, and upload only a single model! (for more details check tf.train.Saver)

## General Instructions
This notebook will be used by you to provide your solution, and by us to both assess your solution and enter your marks. It contains three types of sections:

1. **Setup** Sections: these sections set up code and resources for assessment. **Do not edit, move nor copy these cells**.
2. **Assessment** Sections: these sections are used for both evaluating the output of your code, and for markers to enter their marks. **Do not edit, move, nor copy these cells**.
3. **Task** Sections: these sections require your solutions. They may contain stub code, and you are expected to edit this code. For free text answers simply edit the markdown field.  

**If you edit, move or copy any of the setup, assessments and mark cells, you will be penalised with -20 points**.

Note that you are free to **create additional notebook cells** within a task section. 

Please **do not share** this assignment nor the dataset publicly, by uploading it online, emailing it to friends etc.

## Submission Instructions

To submit your solution:

* Make sure that your solution is fully contained in this notebook. Make sure you do not use any additional files other than your saved model.
* Make sure that your solution runs linearly from start to end (no execution hops). We will run your notebook in that order.
* **Before you submit, make sure your submission is tested on the stat-nlp-book Docker setup to ensure that it does not have any dependencies outside of those that we provide. If your submission fails to adhere to this requirement, you will get 0 points**.
* **If running your notebook produces a trivially fixable error that we spot, we will correct it and penalise you with -20 points. Otherwise you will get 0 points for that solution.**
* **Rename this notebook to your `group_X`** (where `X` is the number of your group), and adhere to the directory structure requirements, if you have not already done so. ** Failure to do so will result in -1 point.**
* Download the notebook in Jupyter via *File -> Download as -> Notebook (.ipynb)*.
* Your submission should be a zip file containing the `group_X` directory, containing `group_X.ipynb` notebook, and the `model` directory with _____
* Upload that file to the Moodle submission site.

## <font color='green'>Setup 1</font>: Load Libraries
This cell loads libraries important for evaluation and assessment of your model. **Do not change, move or copy it.**

In [1]:
%%capture
%load_ext autoreload
%autoreload 2
%matplotlib inline
#! SETUP 1 - DO NOT CHANGE, MOVE NOR COPY
import sys, os
_snlp_book_dir = "../../../../../"
sys.path.append(_snlp_book_dir)
# docker image contains tensorflow 0.10.0rc0. We will support execution of only that version!
import statnlpbook.nn as nn

import tensorflow as tf
import numpy as np

## <font color='green'>Setup 2</font>: Load Training Data

This cell loads the training data. **Do not edit the next cell, nor copy/duplicate it**. Instead refer to the variables in your own code, and slice and dice them as you see fit (but do not change their values). 
For example, no one stops you from introducing, in the corresponding task section, `my_train` and `my_dev` variables that split the data into different folds.   

In [2]:
#! SETUP 2 - DO NOT CHANGE, MOVE NOR COPY
data_path = _snlp_book_dir + "data/nn/"
data_train = nn.load_corpus(data_path + "train.tsv")
data_dev = nn.load_corpus(data_path + "dev.tsv")
assert(len(data_train) == 45502)

### Data Structures

Notice that the data is loaded from tab-separated files. The files are easy to read, and we provide the loading functions that load it into a simple data structure. Feel free to check details of the loading.

The data structure at hand is an array of dictionaries, each containing a `story` and the `order` entry. `story` is a list of strings, and `order` is a list of integer indices:

In [3]:
data_train[0]

{'order': [3, 2, 1, 0, 4],
 'story': ['His parents understood and decided to make a change.',
  'The doctors told his parents it was unhealthy.',
  'Dan was overweight as well.',
  "Dan's parents were overweight.",
  'They got themselves and Dan on a diet.']}

## <font color='blue'>Task 1</font>: Model implementation

Your primary task in this assignment is to implement a model that produces the right order of the sentences in the dataset.

### Preprocessing pipeline

First, we construct a preprocessing pipeline, in our case `pipeline` function which takes care of:
- out-of-vocabulary words
- building a vocabulary (on the train set), and applying the same unaltered vocabulary on other sets (dev and test)
- making sure that the length of input is the same for the train and dev/test sets (for fixed-sized models)

You are free (and encouraged!) to do your own input processing function. Should you experiment with recurrent neural networks, you will find that you will need to do so.

In [4]:
# Change capital letters into lower case letters
#for i in range(len(data_train)):
    #for j in range(5):
        #text=data_train[i]["story"][j]
        #text=text.lower()   
        #data_train[i]["story"][j]=text
#for i in range(len(data_dev)):
    #for j in range(5):
        #text=data_dev[i]["story"][j]
        #text=text.lower()   
        #data_dev[i]["story"][j]=text

In [5]:
#through away the punctuations
for i in range(len(data_train)):
    for j in range(5):
        if data_train[i]["story"][j][-1] == '.':
            data_train[i]["story"][j] = data_train[i]["story"][j][:-1] 
for i in range(len(data_dev)):
    for j in range(5):
        if data_dev[i]["story"][j][-1] == '.':
            data_dev[i]["story"][j] = data_dev[i]["story"][j][:-1]

In [6]:
# convert train set to integer IDs
train_stories, train_orders, vocab = nn.pipeline(data_train)

You need to make sure that the `pipeline` function returns the necessary data for your computational graph feed - the required inputs in this case, as we will call this function to process your dev and test data. If you do not make sure that the same pipeline applied to the train set is applied to other datasets, your model may not work with that data!

In [9]:
# get the length of the longest sentence
max_sent_len = train_stories.shape[2]

# convert dev set to integer IDs, based on the train vocabulary and max_sent_len
dev_stories, dev_orders, _ = nn.pipeline(data_dev, vocab=vocab, max_sent_len_=max_sent_len)

You can take a look at the result of the `pipeline` with the `show_data_instance` function to make sure that your data loaded correctly:

In [10]:
nn.show_data_instance(dev_stories, dev_orders, vocab, 155)

Input:
 Story:
  The manager decided to offer John the job
  During the interview he was very talkative and <OOV>
  He went to the interview very prepared and nicely dressed
  John was excited to have a job interview
  The manager of the company was really impressed by John's comments
 Order:
  [4 2 1 0 3]

Desired story:
  John was excited to have a job interview
  He went to the interview very prepared and nicely dressed
  During the interview he was very talkative and <OOV>
  The manager of the company was really impressed by John's comments
  The manager decided to offer John the job


### Model

The model we provide is a rudimentary, non-optimised model that essentially represents every word in a sentence with a fixed vector, sums these vectors up (per sentence) and puts a softmax at the end which aims to guess the order of sentences independently.

First we define the model parameters:

In [11]:
### MODEL PARAMETERS ###
target_size = 5
vocab_size = len(vocab)
input_size = 10
# n = len(train_stories)
output_size = 5

and then we define the model

In [80]:
tf.reset_default_graph()

In [81]:
### MODEL ###

## PLACEHOLDERS
story = tf.placeholder(tf.int64, [None, None, None], "story")        # [batch_size x 5 x max_length]
order = tf.placeholder(tf.int64, [None, None], "order")              # [batch_size x 5 ]

In [82]:
from tensorflow.python.ops import rnn, rnn_cell
## Build the RNN
n_classes = 5
BATCH_SIZE = 32
#Defining training epochs (iterations)
nEpochs = 5
#Defining the size of RNN
rnn_size = 258
# Word embeddings
initializer = tf.random_uniform_initializer(-0.1, 0.1)
embeddings = tf.get_variable("W", [vocab_size, input_size], initializer=initializer)
def Recurrent_neural_network(story):
    global predict
    batch_size = tf.shape(story)[0]
    sentences = [tf.reshape(x, [batch_size, -1]) for x in tf.split(1, 5, story)]  # 5 times [batch_size x max_length]
    sentences_embedded = [tf.nn.embedding_lookup(embeddings, sentence)   # [batch_size x max_seq_length x input_size]
                      for sentence in sentences]

    hs = [tf.reduce_sum(sentence, 1) for sentence in sentences_embedded] # 5 times [batch_size x input_size]

    #Defining the sigle layer RNN
    RNN_layer = {'weight':tf.Variable(tf.random_normal([rnn_size,n_classes])),\
                     'bias':tf.Variable(tf.random_normal([n_classes]))}

    #5 times [batch_size x word_length]
    basic_lstm_cell = rnn_cell.BasicLSTMCell(rnn_size,forget_bias=0.0)
    GRU_cell = rnn_cell.GRUCell(rnn_size)
    lstm_cell = rnn_cell.LSTMCell(rnn_size,forget_bias=0.0)
    hierarchical_cell = rnn_cell.MultiRNNCell([lstm_cell]*2)

    #stacked_lstm = rnn_cell.MultiRNNCell([basic_lstm_cell] * 2, state_is_tuple=False)

    All_Outputs, All_States = rnn.rnn(lstm_cell,hs,dtype=tf.float32)

    # ouput
    Processed_output = tf.concat(1,All_Outputs)
    Processed_output = tf.reshape(Processed_output,[batch_size,5*rnn_size])
    #Linear-process
    result_flat = tf.contrib.layers.linear(Processed_output, 5 * n_classes)    # [batch_size x 5*n_classes]
    result_output = tf.reshape(result_flat, [-1, 5, n_classes])        # [batch_size x 5 x n_classes]

    #Get the prediction
    #Soft-max could not deal with matrix  --- change deal it one by one
    unpacked_final = [tensor for tensor in tf.unpack(result_output, axis=1)]
    softmaxes_final = [tf.nn.softmax(tensor) for tensor in unpacked_final]
    softmaxed_final = tf.pack(softmaxes_final, axis=1)
    predict = tf.arg_max(softmaxed_final,2)

    final_output = result_output


    return final_output

We built our model, together with the loss and the prediction function, all we are left with now is to build an optimiser on the loss:

### Model training 

We defined the preprocessing pipeline, set the model up, so we can finally train the model

In [83]:
#prediction = Recurrent_neural_network(story)

In [84]:
#print(order)
#print(prediction)

In [85]:
import random
#def train_neural_network(story,order):
prediction = Recurrent_neural_network(story)

lambda_l2=0.001
n = train_stories.shape[0]
loss = tf.reduce_sum(tf.nn.sparse_softmax_cross_entropy_with_logits(prediction, order) + \
                lambda_l2 * tf.nn.l2_loss(embeddings))
#loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(prediction,order))
#prediction = tf.reshape(prediction_flat, [-1, 5, target_size])  
global train_predicted,dev_predicted
#Defining the optisimiser
optimiser = tf.train.AdamOptimizer().minimize(loss)

#train_orders dev_orders
with tf.Session() as sess:
    sess.run(tf.initialize_all_variables())

    #Training Session
    for cEpoch in range(nEpochs):
        current_Epoch_Loss = 0
        #for each epoch we need times to perform stochastic gradient descent 
        y=random.sample(range(n),n)
        for i in range(len(train_stories) // BATCH_SIZE):
            inst_story = train_stories[y[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]]
            #inst_story = inst_story.tolist()
            #inst_order = label_one_hot_train_oders[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]
            inst_order = train_orders[y[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]]
            #inst_order = inst_order.tolist()
            _,currentloss = sess.run([optimiser,loss],feed_dict = {story: inst_story, order: inst_order})
            current_Epoch_Loss += currentloss
        print(cEpoch,'has been completed and the loss of this epoch is',current_Epoch_Loss/train_stories.shape[0])
        train_feed_dict = {story: train_stories, order: train_orders}
        train_predicted = sess.run(predict, feed_dict=train_feed_dict)
        train_accuracy = nn.calculate_accuracy(train_orders, train_predicted)
        print(' Train accuracy:', train_accuracy)

        dev_feed_dict = {story: dev_stories, order: dev_orders}
        dev_predicted = sess.run(predict, feed_dict=dev_feed_dict)
        dev_accuracy = nn.calculate_accuracy(dev_orders, dev_predicted)
        print(' Dev accuracy:', dev_accuracy)

        if dev_accuracy>0.54:
            nn.save_model(sess)

0 has been completed and the loss of this epoch is 6.17049649741
 Train accuracy: 0.555105270098
 Dev accuracy: 0.504649919829
1 has been completed and the loss of this epoch is 5.40709237791
 Train accuracy: 0.603199859347
 Dev accuracy: 0.539176910743
2 has been completed and the loss of this epoch is 5.24023405978
 Train accuracy: 0.614983956749
 Dev accuracy: 0.529877071085
3 has been completed and the loss of this epoch is 5.14401264517
 Train accuracy: 0.62735264384
 Dev accuracy: 0.540673436665
4 has been completed and the loss of this epoch is 5.05820906668
 Train accuracy: 0.637862071997
 Dev accuracy: 0.547942276857


In [86]:
#train_neural_network(story,order)

In [ ]:
# #Alternative way:

#   ## Create a dictionary which enable the sentence order to be corresponded to the status order
# import itertools 
# combination = list(itertools.permutations([0,1,2,3,4])) 
# for i in range(5): 
#     print(combination[i], i)
# ## Check the orders and change them into status representation
# encoded_train_oder = []
# for order in train_orders:
#     for i in range(len(combination)):
#         if np.all(order == combination[i]):
#             encoded_train_oder.append(i)
# ## Check the orders and change them into status representation
# encoded_dev_oder = []
# for order in dev_orders:
#     for i in range(len(combination)):
#         if np.all(order == combination[i]):
#             encoded_dev_oder.append(i) 
    
    
# import random
# story = tf.placeholder(tf.int32, [None, None, None], "story")        # [batch_size x 5 x max_length]
# order = tf.placeholder(tf.int32, [None], "order")              # [batch_size x 5]
# story1_len=tf.placeholder(tf.int32, [None], "story1_len")
# story2_len=tf.placeholder(tf.int32, [None], "story2_len")  
# story3_len=tf.placeholder(tf.int32, [None], "story3_len")  
# story4_len=tf.placeholder(tf.int32, [None], "story4_len")  
# story5_len=tf.placeholder(tf.int32, [None], "story5_len")  
# batch_size = tf.shape(story)[0]


# from tensorflow.python.ops import rnn, rnn_cell
# ## Build the RNN
# n_classes = 120
# BATCH_SIZE = 32
# #Defining training epochs (iterations)
# nEpochs = 10
# #Defining the size of RNN
# rnn_size =120
# # Word embeddings
# initializer = tf.random_uniform_initializer(-0.1, 0.1)
# embeddings = tf.get_variable("W", [vocab_size, input_size], initializer=initializer)
# n_hidden_units=50
def Recurrent_neural_network_full_states(story):
    global predited_endoded_order
    batch_size = tf.shape(story)[0]
    sentences = [tf.reshape(x, [batch_size, -1]) for x in tf.split(1, 5, story)]  # 5 times [batch_size x max_length]
    sentences_embedded = [tf.nn.embedding_lookup(embeddings, sentence)   # [batch_size x max_seq_length x input_size]
                      for sentence in sentences]

    sentences_embedding = [sentence for sentence in sentences_embedded]
    with tf.variable_scope("encoder") as varscope:
        lstm_cell = tf.nn.rnn_cell.LSTMCell(n_hidden_units, state_is_tuple=True)
        _, story1_final_state =tf.nn.dynamic_rnn(lstm_cell, sentences_embedding [0],sequence_length=story1_len,dtype=tf.float32) 
        story1_h = story1_final_state.h
        varscope.reuse_variables()
        _, story2_final_state =tf.nn.dynamic_rnn(lstm_cell, sentences_embedding [1],sequence_length=story2_len,dtype=tf.float32) 
        story2_h = story2_final_state.h
        varscope.reuse_variables()
        _, story3_final_state =tf.nn.dynamic_rnn(lstm_cell, sentences_embedding [2],sequence_length=story3_len,dtype=tf.float32) 
        story3_h = story3_final_state.h
        varscope.reuse_variables()
        _, story4_final_state =tf.nn.dynamic_rnn(lstm_cell, sentences_embedding [3],sequence_length=story4_len,dtype=tf.float32) 
        story4_h = story4_final_state.h
        varscope.reuse_variables()
        _, story5_final_state =tf.nn.dynamic_rnn(lstm_cell, sentences_embedding [4],sequence_length=story5_len,dtype=tf.float32) 
        story5_h = story5_final_state.h
    pair_representation = []
    pair_representation.append(story1_h)
    pair_representation.append(story2_h)
    pair_representation.append(story3_h)
    pair_representation.append(story4_h)
    pair_representation.append(story5_h)
    hs=pair_representation
    h = tf.concat(1, hs)    # [batch_size x 5*input_size]
    h = tf.reshape(h, [batch_size, 5,50])
    #pair_representation = tf.convert_to_tensor(pair_representation)
    #pair_representation=story1_h,story2_h,story3_h,story4_h,story5_h
    #pair_representation=tf.shape(pair_representation,[batch_size,5,-1])
    #Defining the sigle layer RNN
    basic_lstm_cell = rnn_cell.BasicLSTMCell(rnn_size)
    GRU_cell = tf.nn.rnn_cell.GRUCell(rnn_size)
    lstm_cell2 = rnn_cell.LSTMCell(rnn_size)
    hierarchical_cell = rnn_cell.MultiRNNCell([lstm_cell2]*2)
    All_Outputs, All_States = rnn.dynamic_rnn( lstm_cell2 ,h, time_major=False,dtype=tf.float32)

    ################Full state version################
    Processed_output = tf.concat(1,All_Outputs)
    Processed_output = tf.reshape(Processed_output,[batch_size,5*rnn_size])
    ##############Only last output###################
    #Processed_output = tf.concat(1,All_Outputs[:,4,:])
    #Processed_output = tf.reshape(Processed_output,[batch_size,rnn_size])
    # ouput
    final_result = tf.contrib.layers.linear(Processed_output, n_classes)    # [batch_size x n_classes]
    #final_result = tf.add(tf.matmul(All_Outputs[:,4,:],RNN_layer['weight']),RNN_layer['bias'])

    unpached_probability = tf.nn.softmax(final_result) 
    predited_endoded_order = tf.arg_max(unpached_probability,1) 

    final_output = final_result


    return final_output


# import random
def train_neural_network(story,order):
    prediction = Recurrent_neural_network_full_states(story)
    
    lambda_l2=0.001
    n = train_stories.shape[0]
    loss = tf.reduce_sum(tf.nn.sparse_softmax_cross_entropy_with_logits(prediction, order) + \
                    lambda_l2 * tf.nn.l2_loss(embeddings))
    #loss = tf.reduce_mean(tf.nn.sparse_softmax_cross_entropy_with_logits(prediction,order))
    #prediction = tf.reshape(prediction_flat, [-1, 5, target_size]
    #global train_predicted,dev_predicted
    #Defining the optisimiser
    optimiser = tf.train.AdamOptimizer().minimize(loss)
    
    #train_orders dev_orders
    with tf.Session() as sess:
        sess.run(tf.initialize_all_variables())

        #Training Session
        for cEpoch in range(nEpochs):
            current_Epoch_Loss = 0
            #for each epoch we need times to perform stochastic gradient descent 
            for i in range(len(train_stories) // BATCH_SIZE):
                inst_story = train_stories[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]
                #inst_story = inst_story.tolist()
                #inst_order = label_one_hot_train_oders[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]
                inst_order = encoded_train_oder[i * BATCH_SIZE: (i + 1) * BATCH_SIZE]
                #inst_order = inst_order.tolist()
                _,currentloss = sess.run([optimiser,loss],feed_dict  = {story: inst_story, order: inst_order,story1_len:stories_len[i * BATCH_SIZE: (i + 1) * BATCH_SIZE,0],
                         story2_len:stories_len[i * BATCH_SIZE: (i + 1) * BATCH_SIZE,1],story3_len:stories_len[i * BATCH_SIZE: (i + 1) * BATCH_SIZE,2],story4_len:stories_len[i * BATCH_SIZE: (i + 1) * BATCH_SIZE,3],story5_len:stories_len[i * BATCH_SIZE: (i + 1) * BATCH_SIZE,4]})
                current_Epoch_Loss += currentloss
            print(cEpoch,'has been completed and the loss of this epoch is',current_Epoch_Loss/train_stories.shape[0])
            train_feed_dict = {story: train_stories, order:encoded_train_oder,story1_len:stories_len[:,0],story2_len:stories_len[:,1],story3_len:stories_len[:,2],story4_len:stories_len[:,3],story5_len:stories_len[:,4]}
            train_predicted_encoded = sess.run(predited_endoded_order, feed_dict=train_feed_dict)
            train_predicted = []
            for i in range(len(train_predicted_encoded)):
                train_predicted.append(combination[train_predicted_encoded[i]])
            train_accuracy = nn.calculate_accuracy(train_orders, train_predicted)
            print(' Train accuracy:', train_accuracy)

            dev_feed_dict = {story: dev_stories, order:encoded_dev_oder,story1_len:dev_stories_len[:,0],story2_len:dev_stories_len[:,1],story3_len:dev_stories_len[:,2],story4_len:dev_stories_len[:,3],story5_len:dev_stories_len[:,4]}
            dev_predicted_encoded = sess.run(predited_endoded_order, feed_dict=dev_feed_dict)
            dev_predicted = []
            for i in range(len(dev_predicted_encoded)):
                dev_predicted.append(combination[dev_predicted_encoded[i]])
            dev_accuracy = nn.calculate_accuracy(dev_orders, dev_predicted)
            print(' Dev accuracy:', dev_accuracy)

            
# train_neural_network(story,order)

## <font color='red'>Assessment 1</font>: Assess Accuracy (50 pts) 

We assess how well your model performs on an unseen test set. We will look at the accuracy of the predicted sentence order, on sentence level, and will score them as followis:

* 0 - 20 pts: 45% <= accuracy < 50%, linear
* 20 - 40 pts: 50% <= accuracy < 55
* 40 - 70 pts 55 <= accuracy < Best Result, linear

The **linear** mapping maps any accuracy value between the lower and upper bound linearly to a score. For example, if your model's accuracy score is $acc=54.5\%$, then your score is $20 + 20\frac{acc-50}{55-50}$.

The *Best-Result* accuracy is the maximum of the best accuracy the course organiser achieved, and the submitted accuracies scores.  

Change the following lines so that they construct the test set in the same way you constructed the dev set in the code above. We will insert the test set instead of the dev set here. **`test_feed_dict` variable must stay named the same**.

In [87]:
# LOAD THE DATA
data_test = nn.load_corpus(data_path + "dev.tsv")
for i in range(len(data_test)):
    for j in range(5):
        if data_test[i]["story"][j][-1] == '.':
            data_test[i]["story"][j] = data_test[i]["story"][j][:-1] 

# make sure you process this with the same pipeline as you processed your dev set
test_stories, test_orders, _ = nn.pipeline(data_test, vocab=vocab, max_sent_len_=max_sent_len)

# THIS VARIABLE MUST BE NAMED `test_feed_dict`
test_feed_dict = {story: test_stories, order: test_orders}


for i in range(len(data_test)):
    for j in range(5):
        if data_test[i]["story"][j][-1] == '.':
            data_test[i]["story"][j] = data_test[i]["story"][j][:-1]

The following code loads your model, computes accuracy, and exports the result. **DO NOT** change this code.

In [88]:
#! ASSESSMENT 1 - DO NOT CHANGE, MOVE NOR COPY
with tf.Session() as sess:
    # LOAD THE MODEL
    saver = tf.train.Saver()
    saver.restore(sess, './model/model.checkpoint')
    
    # RUN TEST SET EVALUATION
    dev_predicted = sess.run(predict, feed_dict=test_feed_dict)
    dev_accuracy = nn.calculate_accuracy(dev_orders, dev_predicted)

dev_accuracy

0.54794227685729557

## <font color='orange'>Mark</font>:  Your solution to Task 1 is marked with ** __ points**. 
---

## <font color='blue'>Task 2</font>: Describe your Approach

Enter a 750 words max description of your approach **in this cell**.
Make sure to provide:
- an **error analysis** of the types of errors your system makes
- compare your system with the model we provide, focus on differences and draw useful comparations between them

Should you need to include figures in your report, make sure they are Python-generated. For that, feel free to create new cells after this cell (before Assessment 2 cell). Link online images at your risk.


In this assignment, our team tried to improve each step of establishing the model of ordering sentences and using different methods to improve the main model.
 
The first step is to achieve better tokenization, our team use two methods. The first method is to transfer all upper case words into lower case, and substitute the new words into the original training set. Then, the second is to remove the punctuations in each sentences. After testing the model, the transformation of words does not contribute to a higher final result, so first one has been deleted.
 
After processing the original data, we try to improve the word embedding. We try to use Word2vec (word to vector) to pre-train each word, and put the words into one vector with regard to similarity of the main word (one word selected as ‘main’ word in a sentence, and is used to obtain its similarities with its neighbours). Based on that idea, we can have a primary prediction subject to words. However, the result of adding Word2vec does not significantly increase the accuracy of our model. The dimension of each words vector is 10 in this situation. Generally, the dimension should be at least 50. Therefore, the bad result of adding Word2vec may be attributed to the insufficient size of dimensions. In this assignment, the size of all the samples is not enough to create the 50 dimensions words vector.
 
After words being pre-trained, we try to use RNN (Recurrent Neural Network) model to replace the original one. The original one is a linear model, and it cannot capture the information of sentences and the relationship among sentences from a sentence group.
 
Generally, we divide five sentences in each batch (group) into five sets, and assign indices to five sets and transformed each index into vectors. Then, put those vectors into cells to get the outputs. In this step, we use four different cells in the main model, basic LSTM, LSTM, GRU and MultiRNN(hierarchical). After comparing results of each cells, LSTM cell showed the best result. Furthermore, we use linear transformation to reshape the output, softmax to calculate the probability of each output and argmax to choose the highest probability. Finally, we can get the prediction for the sentence order. We implement this process to sentences with repeated orders to build the model.
 
Alternatively, we tried to remove the repeated orders for the whole process. Specifically, one model we tried to use is ‘full sequence encoding’. There are totally 120 possible sequences of five sentences, as 120 is the number of combinations of how ‘0’, ‘1’, ‘2’, ‘3’, ‘4’ can be ordered into different sequences. Every sentence is 50 dimensional trained by the first RNN cell, which forms a 250-dimensional vector for five sentences. Trained after the second RNN cell, it generates five outputs (every output is 120 dimensional) and we combine them into just one output (600 dimensions) which is then trained by a linear model in order to reduce dimensions from 600 to 120. Finally, we use Softmax to get probabilities of all these possible options. The highest probability is the one that the most probable sequence order it would be.
 
In both models, there is still a chance that one label might be decided as another label as a mistake. Furthermore, by comparing those two models, predicting the repeated orders may cause some insignificant outcomes but requires attention as well, so the second one seems more rational and rigorous. Surprisingly, the first one gave us higher accuracy. This result is probably because of a small sample size. With the increment of sample size, the results may differ. Meanwhile, if we regard the problem as a 5-classes classification problem, we could perform precision-recall analysis for the 5 classes. The result could be shown as below:

|label|Precision|Recall|
|:-----|----------------------|----------:|
|0| 0.168|0.225|
|1|0.166|0.223|
|2|0.167|0.165|
|3|0.166|0.166|
|4|0.168|0.225|
 
As we could see, the result is not very good. That's probably because we have both mismatching and repeatative matching problem.

At the end, we try to train the model. We compare the prediction order to the real sentence order to find the prediction loss of the model.  Also, we use L2 regularization to prevent overfitting generated by the prediction process. Also, we use stochastic gradient to take samples from training set to keep the authenticity of each gradient.

the comparison are shown below:


|Model applied|Accuracy|
|:----------------------|----------:|
|Word to Vector + Original Model|51.3% to 51.5%|
|Word to Vector + RNN|about 0.52|
|RNN|about 0.55 (0.556 as optimal)|
|RNN + Remove repition in orders|about 0.53|







## <font color='red'>Assessment 2</font>: Assess Description (30 pts) 

We will mark the description along the following dimensions: 

* Clarity (10pts: very clear, 0pts: we can't figure out what you did, or you did nothing)
* Creativity (10pts: we could not have come up with this, 0pts: Use only the provided model)
* Substance (10pts: implemented complex state-of-the-art classifier, compared it to a simpler model, 0pts: Only use what is already there)

## <font color='orange'>Mark</font>:  Your solution to Task 2 is marked with ** __ points**.
---

## <font color='orange'>Final mark</font>: Your solution to Assignment 3 is marked with ** __points**. 